### PR TITLE
Code: Style add zeebo/errs usage

### PR DESCRIPTION
**Pull request by ifraixedes**
~While doing some reviews, I found some lines where I asked about if they were conventions in Storj, some of them because the style looked odd to me, others where believing that we mostly prefer explicit code than shorter one looked the opposite to that and others because the response was more like something that we apply mechanically without thinking the semantics.~

~The responses haven't been 100% affirmative that they should be in this way, but they could be based on the current code base, hence my way to find out if they are actually conventions or not has been to open this PR and have a review for it.~

~NOTE that each commit contain a specific thing, so if some of them turned to be agreed to be added but not all of them, it should be easy to drop some commit.~

~References on each one added [commit - link]:~

* ~8095055 - https://github.com/storj/storj/pull/1972#discussion_r285583149~
* ~0033cb4 - https://github.com/storj/storj/pull/1972#discussion_r285629766~
* ~30c28bb - https://github.com/storj/storj/pull/1972#discussion_r285090510~

The ones commented here that aren't seen as a convention have been removed and now the PR only contains:

* Mention the package that we use for wrapping and annotating errors, github.com/zeebo/errs.
* Add the case where we avoid errors conditionals when the errors must be only wrapped or annotated by zeebo/errs package.

In order of unblocking these changes of getting merged due to the controversial of the rest.

### CODE REVIEW

File: code/Style.md [hunk 1]
```diff
@@ -70,6 +72,48 @@ func Example() (err error) {
...
+```go
+var ErrUser = errs.Class("user")
+
+func (reg *Registry) GetUser(ctx context.Context, username string) (*User, error) {
+	user, := reg.db.Get_User_By_Username(ctx, username)
+	if err == sql.ErrNoRows {
+		return nil, ErrUser.New("not found %s", username)
+	}
+	if err != nil {
+		return nil, Error.Wrap(err)
```
Comment (VinozzZ) [inline code/Style.md]: What's the reason behind using the generic `Error` here vs `ErrUser`?

Comment (ifraixedes) [reply code/Style.md]: None, it was forgotten to be changed when copied and pasted.
I fixed it in a new commit.

File: code/Style.md [hunk 2]
```diff
@@ -70,6 +72,48 @@ func Example() (err error) {
...
+```go
+func OpenDataFile(filename string) (*os.File, error) {
+	file, err := os.Open(filepath.Join("data", filename))
+	return file, errs.Wrap(err)
+}
+```
+
+### Conditionals for _Not Found_ error values
+
+It's pretty common in Go to find functions that return an error when a function should return a value but such value "isn't found". The most classic example of this is the [`sql.ErrNoRows`](https://golang.org/pkg/database/sql/#pkg-variables) which may be returned by function like [`sql#DB.QueryRow`](https://golang.org/pkg/database/sql/#DB.QueryRow). Under this circumstances we first check for such error and if it doesn't match then we do a second check for non `nil` error, rather than nesting them.
```
Comment (thepaul) [inline code/Style.md]: This seems to be strictly in the realm of aesthetic preference. Does it really matter one way or the other?

File: code/Style.md [hunk 3]
```diff
@@ -197,6 +241,24 @@ Cognitive load at 5 arguments is quite high already, so at that moment prefer co
 
 _Exclude `ctx` and `log` from the count of 5, since common arguments don't have as drastic effect._
 
+## Struct types constructors
+
+Constructor functions of struct types must always return a pointer to the type rather than a value.
```
Comment (thepaul) [inline code/Style.md]: Why must they? It seems to me that there are several situations where it's better to return a value.

Also, implementing this rule would require a precise definition of "constructor function", and I'm not sure what that would look like.

File: code/Style.md (hunk 2 above)
Comment (ifraixedes) [reply code/Style.md]: Not to me, I'm fine with both, it looked odd to me the one written here, but because that one is spread through the sources I wanted to figure it out if it's a convention for aligning my future contributions to such code style.

File: code/Style.md (hunk 3 above)
Comment (ifraixedes) [reply code/Style.md]: > Why must they? It seems to me that there are several situations where it's better to return a value.

I agree with that, I use a reference or a value depending of the case; however based on the response I wanted to know if for avoiding to think if a value fits better than a pointer, the convention is to always use a pointer, because a pointer works everywhere meanwhile a value not.

Another thing is about performance, which is always tricky to know without benchmarking/profiling and there isn't always time to spend for those, hence, I prefer to use a value or type based on semantics when there is not a clear gain and figuring out later on that there is a bottle neck and change it.

You know Go has transitioned from having performance issues with pointers over values but that started to change with the concurrent GC, so there isn't always a clear answer to use one or another without benchmarking/profiling.

> Also, implementing this rule would require a precise definition of "constructor function", and I'm not sure what that would look like.

It's a term used commonly in Go, mostly they are package exported functions which return an instance of a type and their names are commonly prefixed with `New`.

We can add that definition if this is confirmed as a convention and it gets added to the Style.

File: code/Style.md (hunk 2 above)
Comment (egonelbre) [reply code/Style.md]: The nesting thing is because of some of the code relies on checks `err == sql.ErrNoRows`... if we hide the error `sql.ErrNoRows`, then we lose the capability to check it.

However, `errs.Is` should help us fix these situations.

Comment (ifraixedes) [reply code/Style.md]: This is more about the order of the conditionals.

File: code/Style.md (hunk 3 above)
Comment (egonelbre) [reply code/Style.md]: Yeah, I would rather not have a convention around this. I generally agree with the statement, but might be over-arching.

I think the majority comes from the difference whether it's meant to be used as a value or as an object.

File: code/Style.md (hunk 2 above)
Comment (egonelbre) [reply code/Style.md]: Ah, order wise, sometimes we've been using:

``` go
if err != nil {
    if err == sql.ErrNoRows {
        return nil, ErrUser.New("not found %s", username)
    }
    return nil, ErrUser.Wrap(err)
}
```

This is to make clear that when the error happens it's always terminated.

File: code/Style.md (hunk 3 above)
Comment (ifraixedes) [reply code/Style.md]: Good to know that we don't want to have it as a convention, because I tended to use values or pointers depending of the semantics, however the answer that I've got in that comment, was that this probably was a convention.

I gonna remove this one.

Comment (egonelbre) [reply code/Style.md]: It's more because we implement services and similar things and they are usually pointers...

File: code/Style.md (hunk 2 above)
Comment (ifraixedes) [reply code/Style.md]: That's what I use, what I've seen that's commonly used and to me it's more clear for the reason that you wrote, however I found the other order several times in the code and I thought that there was such convention or preference.

Comment (ifraixedes) [reply code/Style.md]: Do we agree that we should use the most common way?

Comment (egonelbre) [reply code/Style.md]: I'm not yet sure either way, I haven't gone through the code-base to see what the motivations might be for the alternate approaches.

**Comment by ifraixedes** [comment]
@egonelbre @thepaul I've updated the description on what has changed in this PR with the goal of getting it merged.

File: code/Style.md [hunk 4]
```diff
@@ -70,6 +72,28 @@ func Example() (err error) {
...
+	return file, nil
+}
+```
+
+should be written like
+
+```go
+func OpenDataFile(filename string) (*os.File, error) {
+	file, err := os.Open(filepath.Join("data", filename))
+	return file, errs.Wrap(err)
```
Comment (egonelbre) [inline code/Style.md]: Shouldn't this be `Error.Wrap`?

File: code/Style.md [hunk 5]
```diff
@@ -70,6 +72,28 @@ func Example() (err error) {
 }
 ```
 
+Sometimes, there are situations where a function, calls another function that only needs to return the value or the error returned by the call; when the called function return errors not created or wrapped by `errs` package the error must be  wrapped or annotated, however it isn't needed to have a conditional to achieve such purpose, because the functions that `errs` package have ignore the `nil` value when they are passed through; for example
```
Comment (egonelbre) [inline code/Style.md]: Let's list exceptions here where we shouldn't use wrapping.

1. Read/Write etc. expect `io.EOF` returned.
2. some packages require specific error codes

File: code/Style.md (hunk 4 above)
Comment (ifraixedes) [reply code/Style.md]: If there is one yes, but then in this snipped we have to add that `Error` definition otherwise it won't be clear.

File: code/Style.md (hunk 5 above)
Comment (ifraixedes) [reply code/Style.md]: I'm aware of those now, this PR was too old and I didn't reread, so good point, I will add them

File: code/Style.md (hunk 4 above)
Comment (ifraixedes) [reply code/Style.md]: oh! I've just seen that I did that in the previous snipped.

Do you think that `Error` definition should be added to both?

Comment (egonelbre) [reply code/Style.md]: Yeah, I think so.

File: code/Style.md [hunk 6]
```diff
@@ -70,6 +72,38 @@ func Example() (err error) {
 }
 ```
 
+There are __2 exceptions__ when an error **NEVER MUST BE** wrapped nor annotated nor combined:
+
+1. Those functions which return the [`io.EOF`](https://golang.org/pkg/io/#pkg-variables) to indicate that they have finished and they don't have to be called again, like the types which implement the [`io.Reader` interface](https://golang.org/pkg/io/#Reader).
+2. Some packages which require that user specific functions return specific error types for acting appropriately, like [_gRPC status](https://godoc.org/google.golang.org/grpc/status) for being able to respond a [specific protocol error status code](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md).
+
```
Comment (thepaul) [inline code/Style.md]: There are more exceptions, unfortunately. Anywhere 3rd party code relies on a particular value in an error, we can't wrap it without breaking things. For example, I recently ran into code that checks `os.IsNotExist` and `os.IsExist` on error values. Those don't work anymore when matching errors are wrapped.

How does one know for sure whether such values might be in an error? In general, we can't, but I suppose we could try to build up a big list of things to check for. I would much, much rather have us come up with something simpler for the developer, though. Maybe an automated checker that notices when a type implements `io.Reader` or `io.Writer` (etc) and complains when errors are wrapped unconditionally? Maybe we use a special different type from `error` when we expect wrapped things? Maybe we use some kind of special annotation on functions whose error return values should not be wrapped unconditionally?

In short, this seems like a big flaw with our current approach.

Comment (ifraixedes) [reply code/Style.md]: I can change the `1` to be more generic; it isn't very shiny as a full list or your ideas, but I think that it's the first step to improve our docs and doing it incrementally, otherwise, we wouldn't have any until we can dedicate resources and time to work on those.

Let me know if that's fine, otherwise, I will close this PR without merging it.

Comment (egonelbre) [reply code/Style.md]: Even if we want to automate these checks, we need to first list out all the situations where we need to preserve the errors.

Comment (ifraixedes) [reply code/Style.md]: hence, should we start with this one and more through the time that we realize?

Comment (egonelbre) [reply code/Style.md]: Yeah, we can start automating later.

Comment (ifraixedes) [reply code/Style.md]: @thepaul I modified, please read again if that's better for you as the first starting point.

**Comment by ifraixedes** [comment]
It has never been approved so I abandon it.
